MobileSdk/src/main/java/com/spruceid/mobile/sdk/GattClient.kt

```diff
@@ -333,101 +347,312 @@ class GattClient(private var callback: GattClientCallback,
-                    callback.onError(Error("Error reading L2CAP characteristic."))
+        while (true) {
+            val buf = ByteArray(L2CAP_BUFFER_SIZE)
+            try {
+                val numBytesRead = inStream.read(buf)
+                if (numBytesRead == -1) {
+                    reportError("Failure reading request, peer disconnected.")
+                    return
                 }
+                payload.write(buf, 0, buf.count())
+
+                dprint("Currently have ${buf.count()} bytes.")
+
+                // We are receiving this data over a stream socket and do not know how large the
+                // message is; there is no framing information provided, the only way we have to
+                // know whether we have the full message is whether any more data comes in after.
+                // To determine this, we take a timestamp, and schedule an event for half a second
+                // later; if nothing has come in the interim, we assume that to be the full
+                // message.
+                //
+                // Technically, we could also attempt to decode the message (it's CBOR-encoded)
+                // to see if it decodes properly.  Unfortunately, this is potentially subject to
+                // false positives; CBOR has several primitives which have unbounded length. For
+                // messages unsing those primitives, the message length is inferred from the
+                // source data length, so if the (incomplete) message end happened to fall on a
+                // primitive boundary (which is quite likely if a higher MTU isn't negotiated) an
+                // incomplete message could "cleanly" decode.
+
+                requestTimestamp = TimeSource.Monotonic.markNow()
+
+                Executors.newSingleThreadScheduledExecutor()
+                    .schedule({
+                                  val curtime = TimeSource.Monotonic.markNow()
+                                  if ((curtime - requestTimestamp) > 500.milliseconds) {
+                                      val message = payload.toByteArray()
+
+                                      reportLog("Request complete: ${message.count()} bytes.")
+                                      callback.onMessageReceived(message)
+                                  }
+                              }, 500, TimeUnit.MILLISECONDS)
+
+            } catch (e: IOException) {
+                reportError("Error on listening input stream from socket L2CAP: ${e}")
                 return
             }
+        }
+    }
 
-            if (!gatt.setCharacteristicNotification(characteristicServer2Client, true)) {
-                callback.onError(Error("Error setting notification on Server2Client."))
+    /**
+     * Thread for writing the response via L2CAP.
+     */
+    fun writeResponse() {
+        val outStream = l2capSocket!!.outputStream
+        try {
+            while (true) {
```

Comment:
```suggestion
            loop {
```

Reply:
That doesn't compile for me, and the only reference I've found to something like that in Kotlin is in a reddit comment that has `loop` as an inline function:

```inline fun loop(body: () -> Unit): Nothing {
    while (true) {
        body()
    }
}```

Is there some other way of accessing `loop {}` ?

Reply:
You're right. For some reason I had thought Kotlin had a `loop`, but reading the documentation, it does not!

Please disregard my comments!